internal/pipeline/pipeline.go

```diff
@@ -164,3 +197,85 @@ type Function struct {
 	ConfigMap map[string]string `yaml:"configMap,omitempty"`
 }
+
+// Validate will validate the function `f`.
+func (f *Function) Validate() error {
+	err := ValidateFunctionName(f.Image)
+	if err != nil {
+		return fmt.Errorf("function name is not valid: %w", err)
+	}
+	// configCnt is used to count the number of configs provided. More than one
+	// config is not allowed.
+	var configCnt = 0
+	if f.ConfigPath != "" {
+		configCnt++
+		if err := ValidatePath(f.ConfigPath); err != nil {
+			return fmt.Errorf("configPath is invalid: %w", err)
+		}
+	}
+	if len(f.ConfigMap) != 0 {
+		configCnt++
+	}
+	if !isNodeZero(&f.Config) {
+		configCnt++
+	}
+	if configCnt > 1 {
+		return fmt.Errorf("only zero or one config is allowed, %d provided", configCnt)
+	}
+
+	return nil
+}
+
+// ValidateFunctionName validates the function name.
+// According to Docker implementation
+// https://github.com/docker/distribution/blob/master/reference/reference.go. A valid
+// name definition is:
+//	name                            := [domain '/'] path-component ['/' path-component]*
+//	domain                          := domain-component ['.' domain-component]* [':' port-number]
+//	domain-component                := /([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])/
+//	port-number                     := /[0-9]+/
+//	path-component                  := alpha-numeric [separator alpha-numeric]*
+// 	alpha-numeric                   := /[a-z0-9]+/
+//	separator                       := /[_.]|__|[-]*/
+func ValidateFunctionName(name string) error {
+	pathComponentRegexp := `(?:[a-z0-9](?:(?:[_.]|__|[-]*)[a-z0-9]+)*)`
+	domainComponentRegexp := `(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])`
+	domainRegexp := fmt.Sprintf(`%s(?:\.%s)*(?:\:[0-9]+)?`, domainComponentRegexp, domainComponentRegexp)
+	nameRegexp := fmt.Sprintf(`^(?:%s\/)?%s(?:\/%s)*$`, domainRegexp,
+		pathComponentRegexp, pathComponentRegexp)
+	matched, err := regexp.MatchString(nameRegexp, name)
+	if err != nil {
+		return err
+	}
+	if !matched {
+		return fmt.Errorf("function name %s isn't valid", name)
+	}
+	return nil
+}
+
+// IsNodeZero returns true if all the public fields in the Node are empty.
+// Which means it's not initialized and should be omitted when marshal.
+// The Node itself has a method IsZero but it is not released
+// in yaml.v3. https://pkg.go.dev/gopkg.in/yaml.v3#Node.IsZero
+// TODO: Use `IsYNodeZero` method from kyaml when kyaml has been updated to
+// >= 0.10.5
+func isNodeZero(n *yaml.Node) bool {
+	return n != nil && n.Kind == 0 && n.Style == 0 && n.Tag == "" && n.Value == "" &&
+		n.Anchor == "" && n.Alias == nil && n.Content == nil &&
+		n.HeadComment == "" && n.LineComment == "" && n.FootComment == "" &&
+		n.Line == 0 && n.Column == 0
+}
+
+// ValidatePath validates `path` and return an error if it's invalid
+func ValidatePath(p string) error {
+	if path.IsAbs(p) {
+		return fmt.Errorf("path is not relative: %s", p)
+	}
+	if path.Clean(p) != p {
```

Comment:
hmm...this will prevent `./` and `./base` to be valid paths which are completely valid. 
This is getting trickier :)

Another way to validate this would be to construct a path such as `path.Join(curr-pkg-path, rel(source-in-the-pipeline)))` and checking if that exist on the filesystem or not. [Now whether or not creates cycle can be caught during the hydration]

Reply:
Yeah, I noticed this. https://github.com/GoogleContainerTools/kpt/pull/1339#discussion_r555357290